src/mpid/ch4/src/ch4_vci.h

```diff
@@ -70,6 +70,142 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_get_vci(int flag, MPIR_Comm * comm_ptr,
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_to_vci(MPIR_Context_id_t context_id)
+{
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id))
+            % MPIDI_global.n_vcis);
+}
+
+/* Map comm and target_rank to vci_idx */
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_rank_to_vci(MPIR_Context_id_t context_id,
+                                                             int target_rank)
+{
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id) + target_rank)
+            % MPIDI_global.n_vcis);
+}
+
+/* Map comm and tag to vci_idx */
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_tag_to_vci(MPIR_Context_id_t context_id, int tag)
+{
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id) + tag)
+            % MPIDI_global.n_vcis);
+}
+
+/* Map comm, target_rank, and tag to vci_idx */
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_rank_tag_to_vci(MPIR_Context_id_t context_id,
+                                                                 int target_rank, int tag)
+{
+    return (MPIDI_mask_int_sign(MPIR_CONTEXT_READ_FIELD(PREFIX, context_id) + target_rank + tag)
+            % MPIDI_global.n_vcis);
+}
+
+/* Return VCI index of a send transmit context.
+ * Used for two purposes:
+ *   1. For the sender side to determine which VCI index of a transmit context
+ *      to send a message from
+ *   2. For the receiver side to determine which VCI of the remote peer to address
+ *      to when sending ack for sync sends
+ * Note: the unused parameters can be used in future
+ *
+ * ctxid_in_effect: communicator's context ID used to calculate the VCI index.
+ *   For an intercommunicator, the right one may be either comm->context_id
+ *   or comm->recvcontext_id, depending on situation.
+ *   This parameter allows caller to explicitly specify context ID.
+ *
+ * When this function is called from the sender side, ctxid_in_effect should be comm->context_id.
+ * Otherwise (receiver side), it should be comm->recvcontext_id.
+ */
+MPL_STATIC_INLINE_PREFIX int MPIDI_get_sender_vci(MPIR_Comm * comm,
+                                                  MPIR_Context_id_t ctxid_in_effect,
+                                                  int sender_rank, int receiver_rank, int tag)
+{
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+    MPIR_Assert(comm);
+    int vci_idx = MPIDI_VCI_INVALID;
+    bool use_user_defined_vci = (comm->hints[MPIR_COMM_HINT_VCI_IDX_SENDER] != MPIDI_VCI_INVALID);
+    bool use_tag = comm->hints[MPIR_COMM_HINT_NO_ANY_TAG];
+
+    /* Compute vci_idx if user did not provide it */
+    if (unlikely(!use_user_defined_vci)) {
+        if (use_tag) {
+            vci_idx = MPIDI_map_contextid_rank_tag_to_vci(ctxid_in_effect, receiver_rank, tag);
+        } else {        /* General unoptimized case */
+            vci_idx = MPIDI_map_contextid_rank_to_vci(ctxid_in_effect, receiver_rank);
+        }
+    } else {    /* Read user provided value */
```

Comment:
Move the comment to a separate line below. The comment applies to the "branch", not the `else` statement.

Reply:
Done